src/main/java/org/opensearch/knn/index/query/ResultUtil.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.search.TotalHits;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.DocIdSetBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+public final class ResultUtil {
+
+    public static final ResultUtil DEFAULT = new ResultUtil();
+
+    /**
+     * Reduce the results to only include the top k results across all leaf results
+     *
+     * @param perLeafResults Results from the list
+     * @param k the number of results across all leaf results to return
+     * @return Reduced results across all leaf results
+     */
+    public List<Map<Integer, Float>> reduceToTopK(List<Map<Integer, Float>> perLeafResults, int k) {
+        Float[] topKScores = new Float[k];
+        Arrays.fill(topKScores, Float.MIN_VALUE);
+
+        // Iterate over all scores to get min competitive score
+        int count = 0;
+        for (Map<Integer, Float> perLeafResult : perLeafResults) {
+            for (Float score : perLeafResult.values()) {
+                if (count < k - 1) {
+                    topKScores[count++] = score;
+                    continue;
+                }
+
+                if (count == k - 1) {
+                    topKScores[count++] = score;
+                    Arrays.sort(topKScores, Comparator.comparingDouble(Float::doubleValue).reversed());
+                    continue;
+                }
+
+                if (score > topKScores[k - 1]) {
+                    topKScores[k - 1] = score;
+                    Arrays.sort(topKScores, Comparator.comparingDouble(Float::doubleValue).reversed());
+                }
+                count++;
+            }
+        }
+
+        // Reduce the results based on min competitive score
+        // List<Map<Integer, Float>> perLeafResultsCopy = perLeafResults.stream().map(HashMap::new).collect(Collectors.toList());
+        perLeafResults.forEach(results -> results.entrySet().removeIf(entry -> entry.getValue() < topKScores[k - 1]));
```

Comment:
> I see that you are removing the docs with score < minCompetitive score. But what will happen if more than 1 doc has == min competitive score, will it create any problem?

In that case, it will return > k results and everything above the min competitive score. This wont be an issue for later functionality. 

> Also, I see that we are iterating over the perLeafResults twice, see if we can reduce it to 1 to save latency.

Problem is, we need first iteration to find the min score and then the next iteration to filter based on min score. There's probably a more complex way to do this that would give better overall time, but I dont see it being very expensive.

Reply:
> In that case, it will return > k results and everything above the min competitive score. This wont be an issue for later functionality.

Good to know. :) 


> > Also, I see that we are iterating over the perLeafResults twice, see if we can reduce it to 1 to save latency.

> Problem is, we need first iteration to find the min score and then the next iteration to filter based on min score. There's probably a more complex way to do this that would give better overall time, but I don't see it being very expensive.

I think if you keep more information related to segment, doc and score in the Heap You can recreate the perLeafResults easily. But its still an optimization that can be taken as the code evolves. I would recommend adding a comment here saying it can be optimized and we will do it.

These can become expensive once we have more segments and large vales of K.